### PR TITLE
Refactoring toxcore/crypto_core

### PR DESCRIPTION
**Pull request by kpp**

### CODE REVIEW

File: toxcore/crypto_core.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+void increment_nonce_number(uint8_t* nonce, uint32_t num)
+{
+    /* NOTE don't use breaks inside this loop
+     * In particular, make sure, as far as possible,
+     * that loop bounds and their potential underflow or overflow
+     * are independent of user-controlled input (you may have heard of the Heartbleed bug).
+     */
+    uint32_t big_endian_num = htonl(num);
+    uint8_t* num_vec = (uint8_t*) &big_endian_num;
```
Comment (0x7CFE) [inline toxcore/crypto_core.cpp]:
``` C++
const uint32_t big_endian_num = htonl(num);
const uint8_t* const num_vec = reinterpret_cast<const uint8_t*>(&big_endian_num);
```

File: toxcore/crypto_core.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    uint8_t num_as_nonce[crypto_box_NONCEBYTES] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,num_vec[0],num_vec[1],num_vec[2],num_vec[3]};
```
Comment (0x7CFE) [inline toxcore/crypto_core.cpp]: It's easy to make a mess with indices and element counts. Better to split init into two parts.

First, initialize the vector with zeroes:

``` C++
uint8_t num_as_nonce[crypto_box_NONCEBYTES] = {0};
```

Then, assign the values (or use deref assignment):

``` C++
num_as_nonce[crypto_box_NONCEBYTES - 4] = num_vec[0];
num_as_nonce[crypto_box_NONCEBYTES - 3] = num_vec[1];
num_as_nonce[crypto_box_NONCEBYTES - 2] = num_vec[2];
num_as_nonce[crypto_box_NONCEBYTES - 1] = num_vec[3];
```

If crypto_box_NONCEBYTES changes someday, this code will stand.

File: toxcore/crypto_core.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    uint32_t i = crypto_box_NONCEBYTES;
+    uint_fast16_t c = 0U;
```
Comment (0x7CFE) [inline toxcore/crypto_core.cpp]: Better use long variable name like `carry`.

File: toxcore/crypto_core.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+void increment_nonce(uint8_t* nonce)
+{
+    /* FIXME use increment_nonce_number(nonce, 1) or sodium_increment (change to little endian)
+     * NOTE don't use breaks inside this loop
+     * In particular, make sure, as far as possible,
+     * that loop bounds and their potential underflow or overflow
+     * are independent of user-controlled input (you may have heard of the Heartbleed bug).
+     */
+    uint32_t i = crypto_box_NONCEBYTES;
+    uint_fast16_t c = 1U;
```
Comment (0x7CFE) [inline toxcore/crypto_core.cpp]: Better use long variable name like `carry`.

File: toxcore/crypto_core.cpp (hunk 2 above)
Comment (kpp) [inline toxcore/crypto_core.cpp]: I can't declare `const uint8_t num_as_nonce` and then change the elements of that array....

Comment (0x7CFE) [inline toxcore/crypto_core.cpp]: Sorry, that was misprint.

File: toxcore/crypto_core.cpp [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+private:
+    Packet_Pointers(); // = delete
+    Packet_Pointers(const Packet_Pointers&); // = delete
+    Packet_Pointers& operator= (const Packet_Pointers&); // = delete
+};
+
+int create_request(const uint8_t* send_public_key, const uint8_t* send_secret_key, uint8_t* packet,
+                   const uint8_t* recv_public_key, const uint8_t* data, uint32_t length, uint8_t request_id)
+{
+    if (MAX_CRYPTO_REQUEST_SIZE < length + 1 + crypto_box_PUBLICKEYBYTES * 2 + crypto_box_NONCEBYTES + 1 + crypto_box_MACBYTES
```
Comment (dkashitsyn) [inline toxcore/crypto_core.cpp:189]: It would be better to simplify the expression by extracting math into constant and use it in condition.

**Pull request merged**